[24.0] Fix datasets API custom keys encoding

davelopez opened the pull request:
Requires #17779
Fixes #17729 since the custom keys will be serialized correctly after this.

This is similar to #17779 but this time addressing history contents.

## TODO
- [x] Add potential missing fields to models
- [x] Increase test coverage for the use of `view` and `keys` query parameters for HDAs and HDCAs.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Add partial models for HDA and HDCA responses

davelopez committed:
Add workaround for pydantic error with UUIDs

davelopez committed:
Remove allow extra model_config in HistoryItemCommon

All fields need to be explicitly declared now for HDA and HDCA responses.

davelopez committed:
Update client API schema

davelopez committed:
Set response_model_exclude_unset for datasets API

Whenever we return AnyHistoryContentItem.

mvdbeek reviewed (approved):
I think that's the right thing to do!

davelopez committed:
Reorder model union from less to more specific

davelopez committed:
Add missing `hashes` field to HDADetailed

davelopez committed:
Add missing DRS ID field to HDADetailed schema

davelopez committed:
Move genome_build field from HDADetailed to HDASummary

Follow up to #15505 that was missing this change.

davelopez committed:
Add response_model_exclude_unset to API endpoints

This time for those endpoints returning HistoryContentsResult

davelopez committed:
Relax URL validation for display apps as it can be relative

davelopez committed:
Add missing `sources` field to HDADetailed schema

davelopez committed:
Move visualizations field from HDADetailed to custom model

Since the visualizations are not strictly part of the detailed view definition.

davelopez committed:
Refactor custom model names and union ordering

davelopez committed:
Remove explicit metadata_* fields from models

As those are dynamic and associated with particular datatypes and not common to all HDAs.

davelopez committed:
Use extra="allow" for custom HDA model

Unfortunately there is no easy way to deal with the dynamic `metadata_*` fields so we need to resort back to this configuration until we can find a better solution for those.

davelopez committed:
Update client API schema

davelopez committed:
Add API test for HDA serialization

mvdbeek reviewed (approved):
Do you want to add a test maybe making sure that the dataset id is not an integer ?

davelopez committed:
Remove duplicated key from view

davelopez committed:
Make sure the ID is encoded in custom models

davelopez committed:
Remove HDCACustom for now

Since getting specific keys from an HDCA seems not implemented yet.

davelopez committed:
Add API tests for collection views

davelopez commented on lib/galaxy/schema/schema.py:
```diff
@@ -116,6 +116,15 @@ class DatasetCollectionPopulatedState(str, Enum):
     FAILED = "failed"  # some problem populating state, won't be populated
 
 
+class HashFunctionNames(str, Enum):
```
This has to be duplicated for now since I couldn't find any common place to put the definition. The other enum is in `lib/galaxy/util/hash_util.py`
